File: sth/accounting_sth/report/jurnal_costing/jurnal_costing.py
# Copyright (c) 2026, DAS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt

# Urutannya sekaligus urutan tampil di report: bengkel dulu karena hasilnya
# dipakai mill dan kebun, sama dengan urutan COSTING_OTOMATIS di
# sth/overrides/accounting_period.py.
COSTING_DOCTYPES = (
	"Costing Bengkel",
	"Costing Mill",
	"Costing Panen",
	"Costing Perawatan",
)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	periode = get_periode(filters)
	dokumen = get_dokumen_costing(periode, filters)
	return get_columns(), get_data(periode, dokumen, filters)


def get_periode(filters):
	"""Accounting Period jadi patokan: company, unit dan rentang tanggalnya
	diambil dari sana, bukan dari filter yang bisa diubah user."""
	if not filters.get("accounting_period"):
		frappe.throw(_("Accounting Period wajib diisi."))

	periode = frappe.db.get_value(
		"Accounting Period",
		filters.accounting_period,
		["name", "company", "unit", "start_date", "end_date"],
		as_dict=True,
	)

	if not periode:
		frappe.throw(_("Accounting Period {0} tidak ditemukan.").format(filters.accounting_period))

	return periode


def get_dokumen_costing(periode, filters):
	"""Costing milik periode ini, per doctype.

	Costing tidak menyimpan link ke Accounting Period; kaitannya lewat
	company + unit + rentang periode yang persis sama, cara yang sama dipakai
	create_costing_on_submit waktu membuatnya.
	"""
	jenis = filters.get("jenis_costing") or COSTING_DOCTYPES
	if isinstance(jenis, str):
		jenis = [jenis]

	dokumen = {}

	for doctype in COSTING_DOCTYPES:
		if doctype not in jenis:
			continue

		costing_filters = {
			"company": periode.company,
			"periode_dari": periode.start_date,
			"periode_sampai": periode.end_date,
			"docstatus": 1,
		}
		if periode.unit:
			costing_filters["unit"] = periode.unit

		nama = frappe.get_all(doctype, filters=costing_filters, pluck="name", order_by="name")
		if nama:
			dokumen[doctype] = nama

	return dokumen


def get_gl_entries(periode, dokumen, filters):
	conditions = []
	values = {"company": periode.company}

	pasangan = []
	for i, (doctype, nama) in enumerate(dokumen.items()):
		key_type = "vt{0}".format(i)
		key_no = "vn{0}".format(i)
		values[key_type] = doctype
		values[key_no] = nama
		pasangan.append(
			"(gle.voucher_type = %({0})s AND gle.voucher_no IN %({1})s)".format(key_type, key_no)
		)

	conditions.append("(" + " OR ".join(pasangan) + ")")

	if filters.get("account"):
		conditions.append("gle.account IN %(account)s")
		values["account"] = filters.get("account")

	if filters.get("cost_center"):
		conditions.append("gle.cost_center IN %(cost_center)s")
		values["cost_center"] = filters.get("cost_center")

	return frappe.db.sql(
		"""
		SELECT gle.voucher_type, gle.voucher_no, gle.posting_date,
			gle.account, gle.cost_center, gle.remarks,
			gle.debit, gle.credit
		FROM `tabGL Entry` gle
		WHERE gle.is_cancelled = 0
			AND gle.company = %(company)s
			AND {conditions}
		ORDER BY gle.voucher_no, gle.account, gle.debit DESC
	""".format(conditions=" AND ".join(conditions)),
		values,
		as_dict=True,
	)


def get_data(periode, dokumen, filters):
	if not dokumen:
		return []

	entries = get_gl_entries(periode, dokumen, filters)

	per_dokumen = {}
	for gle in entries:
		per_dokumen.setdefault(gle.voucher_no, []).append(gle)

	data = []
	grand_debit = 0
	grand_credit = 0

	for doctype in COSTING_DOCTYPES:
		for voucher_no in dokumen.get(doctype, []):
			baris = per_dokumen.get(voucher_no)
			if not baris:
				continue

			total_debit = 0
			total_credit = 0

			for gle in baris:
				total_debit += flt(gle.debit)
				total_credit += flt(gle.credit)
				data.append({
					"voucher_type": gle.voucher_type,
					"voucher_no": gle.voucher_no,
					"posting_date": gle.posting_date,
					"account": gle.account,
					"cost_center": gle.cost_center,
					"keterangan": gle.remarks,
					"debit": flt(gle.debit),
					"credit": flt(gle.credit),
				})

			grand_debit += total_debit
			grand_credit += total_credit

			data.append({
				"voucher_type": doctype,
				"keterangan": _("Total {0}").format(voucher_no),
				"debit": total_debit,
				"credit": total_credit,
				"is_total": 1,
			})

	# Total keseluruhan ditulis sendiri, bukan lewat add_total_row: baris subtotal
	# per dokumen di atas ikut terjumlah kalau add_total_row yang dipakai.
	if data:
		data.append({
			"keterangan": _("Total Seluruh Costing"),
			"debit": grand_debit,
			"credit": grand_credit,
			"is_total": 1,
		})

	return data


def get_columns():
	return [
		{
			"label": _("Jenis Costing"),
			"fieldname": "voucher_type",
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"label": _("No Dokumen"),
			"fieldname": "voucher_no",
			"fieldtype": "Dynamic Link",
			"options": "voucher_type",
			"width": 160,
		},
		{
			"label": _("Tanggal"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width": 100,
		},
		{
			"label": _("Akun"),
			"fieldname": "account",
			"fieldtype": "Link",
			"options": "Account",
			"width": 300,
		},
		{
			"label": _("Cost Center"),
			"fieldname": "cost_center",
			"fieldtype": "Link",
			"options": "Cost Center",
			"width": 240,
		},
		{
			"label": _("Keterangan"),
			"fieldname": "keterangan",
			"fieldtype": "Data",
			"width": 280,
		},
		{
			"label": _("Debit"),
			"fieldname": "debit",
			"fieldtype": "Currency",
			"width": 140,
		},
		{
			"label": _("Kredit"),
			"fieldname": "credit",
			"fieldtype": "Currency",
			"width": 140,
		},
	]

File: sth/accounting_sth/report/jurnal_costing/jurnal_costing.js
// Copyright (c) 2026, DAS and contributors
// For license information, please see license.txt

const JENIS_COSTING = ["Costing Bengkel", "Costing Mill", "Costing Panen", "Costing Perawatan"];

frappe.query_reports["Jurnal Costing"] = {
	filters: [
		{
			fieldname: "accounting_period",
			label: __("Accounting Period"),
			fieldtype: "Link",
			options: "Accounting Period",
			reqd: 1,
			on_change(report) {
				isi_dari_accounting_period(report);
			},
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			read_only: 1,
		},
		{
			fieldname: "unit",
			label: __("Unit"),
			fieldtype: "Link",
			options: "Unit",
			read_only: 1,
		},
		{
			fieldname: "from_date",
			label: __("Periode Dari"),
			fieldtype: "Date",
			read_only: 1,
		},
		{
			fieldname: "to_date",
			label: __("Periode Sampai"),
			fieldtype: "Date",
			read_only: 1,
		},
		{
			fieldname: "jenis_costing",
			label: __("Jenis Costing"),
			fieldtype: "MultiSelectList",
			get_data(txt) {
				return JENIS_COSTING.filter((jenis) =>
					!txt || jenis.toLowerCase().includes(txt.toLowerCase())
				).map((jenis) => ({ value: jenis, description: "" }));
			},
		},
		{
			fieldname: "account",
			label: __("Akun"),
			fieldtype: "MultiSelectList",
			get_data(txt) {
				return frappe.db.get_link_options("Account", txt, {
					company: frappe.query_report.get_filter_value("company"),
				});
			},
		},
		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "MultiSelectList",
			get_data(txt) {
				return frappe.db.get_link_options("Cost Center", txt, {
					company: frappe.query_report.get_filter_value("company"),
				});
			},
		},
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (data && data.is_total) {
			value = `<span style="font-weight: 600">${value}</span>`;
		}

		return value;
	},

	onload(report) {
		isi_dari_accounting_period(report);
	},
};

// Company, unit dan rentang tanggal cuma tampilan — yang dipakai server tetap
// Accounting Period-nya, jadi tiga filter itu selalu ikut periode yang dipilih.
function isi_dari_accounting_period(report) {
	const accounting_period = frappe.query_report.get_filter_value("accounting_period");

	if (!accounting_period) {
		frappe.query_report.set_filter_value("company", "");
		frappe.query_report.set_filter_value("unit", "");
		frappe.query_report.set_filter_value("from_date", "");
		frappe.query_report.set_filter_value("to_date", "");
		report.page.clear_indicator();
		return;
	}

	frappe.db
		.get_value("Accounting Period", accounting_period, ["company", "unit", "start_date", "end_date"])
		.then(({ message }) => {
			if (!message) return;
			frappe.query_report.set_filter_value("company", message.company || "");
			frappe.query_report.set_filter_value("unit", message.unit || "");
			frappe.query_report.set_filter_value("from_date", message.start_date || "");
			frappe.query_report.set_filter_value("to_date", message.end_date || "");
			report.page.set_indicator(__("Periode: {0}", [accounting_period]), "blue");
		});
}
